apps/sync/js/adapters/bookmarks.js

```diff
@@ -0,0 +1,325 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+  Bookmarks Data Adapter
+
+    The feature of Bookmarks adapter is to retrieve Bookmarks records from
+  FxSync Service and merge to Bookmarks DataStore ('firefox-sync-bookmarks')
+  for FxSync only which has the same functionality, and the adapter is based on
+  SyncEngine's design to handle 'update' and 'handleConflict' request. Bookmarks
+  Data Adapter provides Read-Only version currently, and the basic ideas are
+  included:
+  - Traverse all records from Bookmarks Collection (BC).
+  - Add a new or existed bookmark to Bookmarks DataStore (BDS).
+    - Convert the records from BC to BDS format.
+    - Merge a new record and the existed one then write to BDS.
+  - Remove an existed bookmark when receiving a "deleted: true" record.
+**/
```

Comment:
"The purpose of the Bookmarks adapter is to retrieve bookmark records from Firefox Sync and merge them in a local bookmarks DataStore ('firefox-sync-bookmarks') which will be hosting Firefox Sync records only. This adapter implements the API defined by SyncEngine. For now the implementation allows only to retrieve bookmarks from Sync but it is not ready for pushing any changes. The high level algorithm implemented here is the following:
* Traverse all the records coming from Firefox Sync and stored locally in Kinto.
* Convert any new bookmark from the Firefox Sync format to the 'firefox-sync-bookmarks' DataStore one.
* For existing bookmarks merge the remote with the local bookmark after converting them to the proper 'firefox-sync-bookmarks' DS format.
* Remove any existing bookmark that has been remotely deleted (these records come with a 'deleted: true' property)."